jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushLayersStep.java

```diff
@@ -76,34 +75,40 @@
   public ImmutableList<AsyncStep<PushBlobStep>> call() throws ExecutionException {
     try (TimerEventDispatcher ignored =
         new TimerEventDispatcher(buildConfiguration.getEventDispatcher(), DESCRIPTION)) {
-      ImmutableList<? extends AsyncStep<? extends CachedLayer>> cachedLayer =
+      ImmutableList<? extends AsyncStep<? extends CachedLayer>> cachedLayers =
           NonBlockingSteps.get(cachedLayerStep);
 
-      // Constructs a PushBlobStep for each layer.
-      ImmutableList.Builder<AsyncStep<PushBlobStep>> pushBlobStepsBuilder = ImmutableList.builder();
-      for (AsyncStep<? extends CachedLayer> cachedLayerStep : cachedLayer) {
-        ListenableFuture<PushBlobStep> pushBlobStepFuture =
-            Futures.whenAllSucceed(cachedLayerStep.getFuture())
-                .call(() -> makePushBlobStep(cachedLayerStep), listeningExecutorService);
-        pushBlobStepsBuilder.add(() -> pushBlobStepFuture);
+      try (ProgressEventDispatcher progressEventDispatcher =
+          progressEventDispatcherFactory.create("push layers", cachedLayers.size())) {
+        // Constructs a PushBlobStep for each layer.
+        ImmutableList.Builder<AsyncStep<PushBlobStep>> pushBlobStepsBuilder =
+            ImmutableList.builder();
+        for (AsyncStep<? extends CachedLayer> cachedLayerStep : cachedLayers) {
+          ProgressEventDispatcher.Factory childProgressEventDispatcherFactory =
+              progressEventDispatcher.newChildProducer();
+          ListenableFuture<PushBlobStep> pushBlobStepFuture =
+              Futures.whenAllSucceed(cachedLayerStep.getFuture())
+                  .call(
+                      () -> makePushBlobStep(cachedLayerStep, childProgressEventDispatcherFactory),
+                      listeningExecutorService);
+          pushBlobStepsBuilder.add(() -> pushBlobStepFuture);
+        }
```

Comment:
I *think* this `call()` just returns ` list of `AsyncSteps`, which may not have been completed. Does it make sense we emit a completion event at this point when getting out of the `try` block?

Reply:
Getting out the `try` block doesn't emit a "completion event" but rather just sends a `ProgressEvent` with the rest of the `remainingAllocationUnits`. In this case, it would not send anything because `remainingAllocationUnits` is 0 (all taken up by the child producers).